File: 762.prime-number-of-set-bits-in-binary-representation.cpp
/*
 * @lc app=leetcode id=762 lang=cpp
 *
 * [762] Prime Number of Set Bits in Binary Representation
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int countPrimeSetBits(int L, int R) {
    int count = 0;
    for (int i = L; i <= R; i++) {
      if (isPrime(countBit(i))) {
        count++;
      }
    }
    return count;
  }
  int countBit(int n) {
    int count = 0;
    while (n) {
      count += n % 2;
      n >>= 1;
    }
    return count;
  }
  bool isPrime(int n) {
    if (n <= 1) {
      return false;
    }
    int sqrtn = sqrt(n);
    for (int i = 2; i <= sqrtn; i++) {
      if (n % i == 0) {
        return false;
      }
    }
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.isPrime(1), false);
  ASSERT_EQ(s.isPrime(2), true);
  ASSERT_EQ(s.isPrime(3), true);
  ASSERT_EQ(s.isPrime(4), false);
  ASSERT_EQ(s.countBit(6), 2);
  ASSERT_EQ(s.countBit(7), 3);
  ASSERT_EQ(s.countBit(8), 1);
  ASSERT_EQ(s.countBit(9), 2);
  ASSERT_EQ(s.countBit(10), 2);
  ASSERT_EQ(s.countBit(11), 3);
  ASSERT_EQ(s.countBit(12), 2);
  ASSERT_EQ(s.countBit(13), 3);
  ASSERT_EQ(s.countBit(14), 3);
  ASSERT_EQ(s.countBit(15), 4);
  ASSERT_EQ(s.countBit(16), 1);
  ASSERT_EQ(s.countPrimeSetBits(6, 6), 1);
  ASSERT_EQ(s.countPrimeSetBits(15, 15), 0);
  ASSERT_EQ(s.countPrimeSetBits(6, 10), 4);
  ASSERT_EQ(s.countPrimeSetBits(10, 15), 5);
  return 0;
}

File: 762.prime-number-of-set-bits-in-binary-representation.rs
/*
 * @lc app=leetcode id=762 lang=rust
 *
 * [762] Prime Number of Set Bits in Binary Representation
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn count_prime_set_bits(left: i32, right: i32) -> i32 {
        let mut count = 0;
        for i in left..=right {
            if Solution::is_prime(i.count_ones() as i32) {
                count += 1;
            }
        }
        count
    }

    fn is_prime(n: i32) -> bool {
        if n <= 1 {
            return false;
        }
        let sqrtn: i32 = (n as f32).sqrt() as i32;
        for i in 2..=sqrtn {
            if n % i == 0 {
                return false;
            }
        }
        true
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::count_prime_set_bits(6, 10), 4);
}

#[test]
fn test2() {
    assert_eq!(Solution::count_prime_set_bits(10, 15), 5);
}
